test/extensions/filters/network/http_connection_manager/config_filter_dependencies_test.cc

```diff
@@ -0,0 +1,186 @@
+}
+
+// PantryFilter provides a potato, which is not required by any other filter.
+TEST_F(HttpConnectionManagerConfigTest, UnusedProvidencyOk) {
+  auto hcm_config = parseHttpConnectionManagerFromYaml(ConfigTemplate);
+  hcm_config.add_http_filters()->set_name("test.pantry");
+  hcm_config.add_http_filters()->set_name("envoy.filters.http.router");
+
+  PantryFilterFactory pf;
+  Registry::InjectFactory<NamedHttpFilterConfigFactory> rf(pf);
+
+  HttpConnectionManagerConfig(hcm_config, context_, date_provider_, route_config_provider_manager_,
+                              scoped_routes_config_provider_manager_, http_tracer_manager_,
+                              filter_config_provider_manager_);
+}
+
+// ChefFilter requires a potato, but no filter provides it.
+TEST_F(HttpConnectionManagerConfigTest, UnmetDependencyError) {
+  auto hcm_config = parseHttpConnectionManagerFromYaml(ConfigTemplate);
+  hcm_config.add_http_filters()->set_name("test.chef");
+  hcm_config.add_http_filters()->set_name("envoy.filters.http.router");
+
+  ChefFilterFactory cf;
+  Registry::InjectFactory<NamedHttpFilterConfigFactory> rc(cf);
+
+  EXPECT_THROW_WITH_MESSAGE(
+      HttpConnectionManagerConfig(hcm_config, context_, date_provider_,
+                                  route_config_provider_manager_,
+                                  scoped_routes_config_provider_manager_, http_tracer_manager_,
+                                  filter_config_provider_manager_),
+      EnvoyException,
+      "Dependency violation: filter 'test.chef' requires a FILTER_STATE_KEY named 'potato'");
+}
+
+// ChefFilter requires a potato, but no preceding filter provides it.
+TEST_F(HttpConnectionManagerConfigTest, MisorderedDependenciesError) {
+  auto hcm_config = parseHttpConnectionManagerFromYaml(ConfigTemplate);
+  hcm_config.add_http_filters()->set_name("test.chef");
+  hcm_config.add_http_filters()->set_name("test.pantry");
+  hcm_config.add_http_filters()->set_name("envoy.filters.http.router");
+
+  // Registration order does not matter.
+  PantryFilterFactory pf;
+  Registry::InjectFactory<NamedHttpFilterConfigFactory> rf(pf);
+  ChefFilterFactory cf;
+  Registry::InjectFactory<NamedHttpFilterConfigFactory> rc(cf);
+
+  EXPECT_THROW_WITH_MESSAGE(
+      HttpConnectionManagerConfig(hcm_config, context_, date_provider_,
+                                  route_config_provider_manager_,
+                                  scoped_routes_config_provider_manager_, http_tracer_manager_,
+                                  filter_config_provider_manager_),
+      EnvoyException,
+      "Dependency violation: filter 'test.chef' requires a FILTER_STATE_KEY named 'potato'");
+}
+
+TEST_F(HttpConnectionManagerConfigTest, UpgradeUnmetDependencyError) {
+  auto hcm_config = parseHttpConnectionManagerFromYaml(ConfigTemplate);
+  auto upgrade_config = hcm_config.add_upgrade_configs();
+  upgrade_config->set_upgrade_type("websocket");
+
+  upgrade_config->add_filters()->set_name("test.chef");
```

Comment:
Can you add 2 upgrade configs, where upgrade1 provides a potato and upgrade2 requires one?
On first read I thought you had your processDeps in the wrong place and while I think you're good, it'd be a good thing to have tested.  Ditto potato in main filter config and consumer in upgrade config, just to regressionproof that they're isolated.

Reply:
Done.